File: problems/33-quickselect/001-kth-largest-element/python/solution.py
def find_kth_largest(nums: list[int], k: int) -> int:
    target = len(nums) - k
    left, right = 0, len(nums) - 1

    while left <= right:
        pivot = partition(nums, left, right)
        if pivot == target:
            return nums[pivot]
        if pivot < target:
            left = pivot + 1
        else:
            right = pivot - 1
    raise ValueError("k is out of range")


def partition(nums: list[int], left: int, right: int) -> int:
    pivot_value = nums[right]
    store = left
    for i in range(left, right):
        if nums[i] <= pivot_value:
            nums[i], nums[store] = nums[store], nums[i]
            store += 1
    nums[store], nums[right] = nums[right], nums[store]
    return store

File: problems/33-quickselect/001-kth-largest-element/java/Solution.java
public class Solution {
    public int findKthLargest(int[] nums, int k) {
        int target = nums.length - k;
        int left = 0, right = nums.length - 1;

        while (left <= right) {
            int pivot = partition(nums, left, right);
            if (pivot == target) return nums[pivot];
            if (pivot < target) left = pivot + 1;
            else right = pivot - 1;
        }
        return -1;
    }

    private int partition(int[] nums, int left, int right) {
        int pivotValue = nums[right];
        int store = left;
        for (int i = left; i < right; i++) {
            if (nums[i] <= pivotValue) {
                swap(nums, i, store++);
            }
        }
        swap(nums, store, right);
        return store;
    }

    private void swap(int[] nums, int a, int b) {
        int temp = nums[a]; nums[a] = nums[b]; nums[b] = temp;
    }
}
